Add standard exports into the documentation

SimenB opened the pull request:
It seems weird to me that this list lives on an open Google Docs [document](https://docs.google.com/document/d/1Q0MXWdwp1mdXCzNRak6bW5LLVylVRXhdi7_21Sg15xQ/edit).

Even if it's work in progress, it should rather be marked as such instead of living somewhere else. This way, you have more control over it, and it makes it more official.
A benefit, other than the obvious standardizing, is discoverability. It took me quite some digging to find this, having Google index it might help.

beorn7 commented:
I guess the reason why those are still in the Google doc is that they are not quite finalized yet.

But they should, obviously, so we could use this PR to do so.

I'll leave a few comments, but I guess @brian-brazil has the strongest opinions on this.

beorn7 commented on content/docs/instrumenting/writing_clientlibs.md:
```diff
@@ -327,12 +327,7 @@ readable formats) if it can be implemented without a significant resource cost.
-In addition, client libraries are ENCOURAGED to also offer whatever makes sense
-in terms of metrics for their language’s runtime (e.g. Garbage collection
-stats).
+at [standard exports](/docs/instrumenting/standard_exports).
```
I'd vote for including it right here instead of having yet another section.

The client library guide is anyway long, so no harm in making it a bit longer.

grobie replied:
SGTM.

SimenB replied:
This was mostly to separate it from clients.

If some things are standardized, it's also relevant for scrapers and consoles, not just for the ones writing clients.

Say I wanna write a generic template for metric visualizations, it'd be nice to know what some standard metrics are named, but I don't care about what conventions for other metrics are, I just want to be able to scrape some default ones.

beorn7 commented on content/docs/instrumenting/standard_exports.md:
```diff
@@ -0,0 +1,35 @@
+in terms of metrics for their language’s runtime (e.g. Garbage collection
+stats).
+
+These exports should have the prefix `process_`. If a language or runtime
```
We need more clarity here. The unified process metrics are prefixed `process_` but the language specific metrics are prefixed `<language>_` or similar (e.g. `go_goroutines`, but note that JVM metrics are prefixed `hotspot_`).

SimenB replied:
I agree!
Ideas on how to formulate that?

brian-brazil replied:
The wording here seems fine to me, as we're just talking about the standard exports - not runtime metrics.

grobie replied:
The current wording above make it sound like standard exports include runtime metrics. It might make sense to improve the wording of the definition then.

beorn7 commented on content/docs/instrumenting/standard_exports.md:
```diff
@@ -0,0 +1,35 @@
+
+## Should
+
+| Metric name                     | Description                                            | Format  |
```
I'd replace `Description` by `Help string` and make the help string given here mandatory. (There is still some stalled work lying around to unify the help strings. We can as well do it right now. But that requires to check where we have already consensus, and where we have deviations.)

SimenB replied:
Deffo! Then description can maybe be more about what it is, in case it's not clear (I had to google the different memory stuff for instance).

I copied the descriptions here from the Go implementation (first one I found, but java seems to have the same)

brian-brazil replied:
Unit, not format - but it's already implicit from the name

beorn7 commented on content/docs/instrumenting/standard_exports.md:
```diff
@@ -0,0 +1,35 @@
+
+## May
+
+TODO: IDK the names of GC ones...
```
That's so language dependent that I wouldn't even try to unify it. Modulo the right prefix, see above.

SimenB replied:
Agreeing on a naming convention could still be nice though, or at least provide some sort of baseline for naming. Might even be enough to say "Java called it this, and Go called it that". Just provide some structure, to keep it more the same across implementations.

brian-brazil replied:
This doesn't belong here as these aren't consistent across languages let alone runtimes.

grobie commented on content/docs/instrumenting/writing_clientlibs.md:
```diff
@@ -327,12 +327,7 @@ readable formats) if it can be implemented without a significant resource cost.
-
-In addition, client libraries are ENCOURAGED to also offer whatever makes sense
-in terms of metrics for their language’s runtime (e.g. Garbage collection
-stats).
```
What's the reason to remove this paragraph? We could even explicitly link to client_golang's go_* metrics/go collector.

beorn7 replied:
The reason is that it is also contained in the export description.

But see my suggestion to inline the whole standard export description right here.

grobie replied:
Github needs to get better with showing comments in real-time. Makes all sense.

SimenB replied:
Yeah, it's moved over to the new article

brian-brazil replied:
I'd inline here. The client library and exporter guides are already a bit off to have in the instrumentation section.

grobie commented:
:+1: Thank you!

All our documents were on Google once (client library guide lines, exporter, exposition formats, etc.). We simply haven't ported this yet, but the presented process metrics were the ones we agreed upon end of 2014.

SimenB commented:
Ok, updated.

Changes made (in separate commit) is:

* Inlined it back into the original file
* Changed the 2 headings as suggested

Did not really change any phrasing, as I think you probably can come up with better ones anyway 😄

brian-brazil commented on content/docs/instrumenting/writing_clientlibs.md:
```diff
@@ -327,17 +327,32 @@ readable formats) if it can be implemented without a significant resource cost.
+
+In addition, client libraries are ENCOURAGED to also offer whatever makes sense
+in terms of metrics for their language’s runtime (e.g. Garbage collection
+stats).
```
I'd append "with an appropriate prefix such as go_, hostspot_" to separate them from the standard exports.

SimenB commented:
@brian-brazil Done 😄

brian-brazil commented on content/docs/instrumenting/writing_clientlibs.md:
```diff
@@ -327,17 +327,32 @@ readable formats) if it can be implemented without a significant resource cost.
+| Metric name                     | Help string                                            | Unit    |
+| ------------------------------- | ------------------------------------------------------ | ------- |
+| `process_cpu_seconds_total`     | Total user and system CPU time spent in seconds.       | Seconds |
+| `process_open_fds`              | Number of open file descriptors.                       | Integer |
```
The unit of open file descriptors is file descriptors.

SimenB commented:
@beorn7 @grobie @brian-brazil any other comments? It'd be great to land this so we can point to it 😄 
Or if stuff is basically set, just some tweaking necessary (as in, the current metrics are fine, none will be added/removed) that's fine as well.

beorn7 commented on content/docs/instrumenting/writing_clientlibs.md:
```diff
@@ -327,17 +327,32 @@ readable formats) if it can be implemented without a significant resource cost.
+| `process_start_time_seconds`    | Start time of the process since unix epoch in seconds. | Seconds          |
+
+In addition, client libraries are ENCOURAGED to also offer whatever makes sense
+in terms of metrics for their language’s runtime (e.g. Garbage collection
```
Garbage -> garbage

beorn7 commented on content/docs/instrumenting/writing_clientlibs.md:
```diff
@@ -327,17 +327,32 @@ readable formats) if it can be implemented without a significant resource cost.
+| ------------------------------- | ------------------------------------------------------ | ---------------  |
+| `process_cpu_seconds_total`     | Total user and system CPU time spent in seconds.       | Seconds          |
+| `process_open_fds`              | Number of open file descriptors.                       | File Descriptors |
+| `process_max_fds`               | Maximum number of open file descriptors.               | Integer          |
```
Unit here should also be "File Descriptors".

On a more general note, I'd write all the units with lower case, not just the bytes.

beorn7 commented:
Just nits left from my side.

We can iterate on larger additions or improvements later. For now, it's great to have this in the real docs.

SimenB commented:
@beorn7 sounds great! I'll fix it as soon as I get to my laptop (half hour). Want me to squash, or do you use GH's squash?

juliusv commented:
@SimenB Thanks! And that was my first time trying out GitHub's squash feature!
